### PR TITLE
CI: remove dependency on cabal-plan (#8891)

### PR DESCRIPTION
**Pull request by ulysses4ever**
Let's see what CI thinks...

### CODE REVIEW

File: validate.sh [hunk 1]
```diff
@@ -369,7 +368,6 @@ step_time_summary() {
 step_build() {
 print_header "build"
 timed $CABALNEWBUILD $TARGETS --dry-run || exit 1
-$CABALPLAN topo --builddir=$BUILDDIR || exit 1
```
Comment (fgaz) [inline validate.sh:372]: If we still want a list of packages (though not topologically sorted) we can do it with a jq command:

```sh
jq -r '."install-plan" | map(."pkg-name" + "-" + ."pkg-version" + " " + ."component-name") | join("\n")' "$BUILDDIR/cache/plan.json"
```

**Comment by ulysses4ever** [comment]
Hmm, CI fails with rather mysterious solver error in the testsuite... https://github.com/haskell/cabal/actions/runs/4598996409/jobs/8123811204?pr=8893

**Comment by fgaz** [comment]
> `>>> -j2 --hide-successes --with-ghc=ghc`

Probably `cabal list-bin` returned nothing in `$(cabal list-bin) -j2 [...]`

File: validate.sh (hunk 1 above)
Comment (ulysses4ever) [reply validate.sh:372]: TBH I don't see a big value in this list. If someone feels strongly about it, I could add this oneliner, perhaps. But I'm not a fan of this solution either: it looks rather fragile to my uniformed taste.

Comment (ulysses4ever) [reply validate.sh:372]: Ah, I guess, it was requested here: https://github.com/haskell/cabal/issues/8440 so there are users for it...

**Comment by ulysses4ever** [comment]
> Probably cabal list-bin returned nothing in $(cabal list-bin some_exe) -j2 [...]

Well, yes, but the solver error I was referring to above happens before this. And because of this error `cabal list-bin` is not functional, as it requires a resolved plan...

**Comment by ulysses4ever** [comment]
All right, i can reproduce it under GHC 844 on Linux but not 944. CI also showed the failure only under 844/Linux and stopped in all other configuration. Still no idea why it happens. The actual build ends fine but then list-bin kicks the solver again for some mysterious reason, and that fails...

**Comment by ulysses4ever** [comment]
Got it: `list-bin` needed to have the same project file as was used for the build.

File: validate.sh (hunk 1 above)
Comment (fgaz) [reply validate.sh:372]: >  One difference between cabal build --dry-run and cabal-plan topo that I noticed today (and you can check it using the CI link above) is that the latter seems to print nearly twice as much as the former. I'm not sure why that is...

`cabal build` does not print cached stuff

**Comment by ulysses4ever** [comment]
All right, please review!

File: validate.sh (hunk 1 above)
Comment (ulysses4ever) [reply validate.sh:372]: Good to know, thank you!

I don't see much emotion one way or another, so I'm just leaving the current version (with your oneliner) for the review...

Comment (andreasabel) [reply validate.sh:372]: Where is the  one-liner replacement for the `cabal-plan` printing?  I cannot see it, I only see the deletion of the `cabal-plan` step.

**Review by andreasabel — CHANGES_REQUESTED** [review]
Thanks for your work on this, @ulysses4ever!

The printing of the complete build plan was added for a reason, which can be read up upon in https://github.com/haskell/cabal/pull/8398#issuecomment-1231935643: Folks relax an upper bound, which is not picked up by the solver due to another dependency, and then CI/build passes and people think relaxing the bound was fine.
So, the correct workflow is: 
1. relax the bound of a dependency,
2. build, 
3. check that the new version of the dependency was indeed picked up.

The inclusion of `cabal-plan` served to enable step 3 by inspecting logs of CI runs.
@andreabedini added this rather recently (Aug 2022) upon my request.  We have to avoid the whack-a-mole-situation, that one developer adds a feature that the next developer removes, which is then missed and added again etc.

If you want to remove it, you need to explain how step 3 of the bump workflow can be facilitated.  Ideally would be a check that in the latest GHC, all the maximum versions of the dependencies are indeed picked up.  Atm, `cabal` does not offer such a check.
As pointed out by @fgaz, `cabal build` does not output the complete build plan, only what has to be built, which does not include existing builds that can be reused.

File: validate.sh (hunk 1 above)
Comment (ulysses4ever) [reply validate.sh:372]: @andreasabel oops: forgot to commit. Fixed.

**Comment by ulysses4ever** [comment]
@andreasabel thanks for the review and I whole-heartedly agree. Let's leave the full list of dependencies then (getting it one way or another). I'm sad to say: I have no clue about the procedure for bumping dependencies you're describing. Perhaps, it'd be good to document it somewhere. Although, I'm not sure which place is the best for such information. Maybe we could have a little "CI Maintainer Guide" or something and reference it from validate.sh and validate.yml...

**Comment by ulysses4ever** [comment]
@fgaz @andreabedini any interest here for a review or an opinion?

File: .github/workflows/validate.yml [hunk 2]
```diff
@@ -132,7 +123,7 @@ jobs:
       - name: Tar cabal head executable
         if: matrix.ghc == env.GHC_FOR_RELEASE
         run: |
-          CABAL_EXEC=$(cabal-plan list-bin --builddir=dist-newstyle-validate-ghc-${{ matrix.ghc }} cabal-install:exe:cabal)
+          CABAL_EXEC=$(cabal list-bin --builddir=dist-newstyle-validate-ghc-${{ matrix.ghc }} --project-file=cabal.project.validate cabal-install:exe:cabal)
           # We have to tar the executable to preserve executable permissions
           # see https://github.com/actions/upload-artifact/issues/38
           if [[ ${{ runner.os }} == 'Windows' ]]; then
```
Comment (fgaz) [inline .github/workflows/validate.yml:129]: `cabal-plan` on the line below. just in case, try grepping for `cabal-plan` so we don't miss any

Comment (ulysses4ever) [reply .github/workflows/validate.yml:129]: Done. I removed it in validate.{sh,yml} and Makefile, and i didn't touch couple other places, which don't seem to be related to CI necessarily. There're Dockerfiles but I don't know what to do about them: i have a feeling they have bitrotted. Anyway, I'd prefer to not get involved with them.

**Review by fgaz — APPROVED** [review]
Thanks!

**Pull request merged**